Treas 17

Milszym opened the pull request.

aolszewska commented on backend/src/ideas/dto/createIdeaDto.ts:
```diff
@@ -1,11 +1,42 @@
+    @ArrayMinSize(1)
+    networks!: IdeaNetworkDto[]
+
+    @ApiPropertyOptional()
```
I'm not sure, but I think by default the fields are required, we need to decorate every optional field with `IsOptional` to validate the fields correctly.

Milszym replied:
I agree, I have added it :)

aolszewska commented on backend/src/ideas/dto/createIdeaDto.ts:
```diff
@@ -1,11 +1,42 @@
+    })
+    @IsNotEmpty()
+    @ArrayMinSize(1)
+    networks!: IdeaNetworkDto[]
```
To validate the objects within an array we need to add
```
@IsArray()
@ValidateNested({each: true})
@Type(() => IdeaNetworkDto)
```

Milszym replied:
Ah you're right. I even forgot to add some tests for it. I have changed this and added tests as well

aolszewska commented on backend/src/ideas/dto/createIdeaDto.ts:
```diff
@@ -1,11 +1,42 @@
-interface CreateIdeaDto {
```
A general comment on the dto files - I think I have introduced a bad naming convention :) the DTO files should have `.dto` suffix, so in this case, we should have `createIdea.dto.ts`.

aolszewska commented on backend/src/ideas/dto/ideaDto.ts:
```diff
@@ -0,0 +1,81 @@
+    id!: string
+
+    @ApiProperty()
+    @IsNotEmpty()
```
As this is a response object, we don't need any validation on it - only the swagger decorators.

Milszym replied:
Yeah you're right, I have removed them

aolszewska commented on backend/src/ideas/ideas.controller.spec.ts:
```diff
@@ -144,12 +199,96 @@ describe(`/api/v1/ideas`, () => {
 
-        it('should return bad request for empty title', async () => {
+    describe('PATCH', () => {
+        it('should return status ok for minima patch data', async () => {
```
Typo, should be`minimal` :)

Milszym replied:
Thanks :) fixed

aolszewska commented on backend/src/ideas/ideas.controller.ts:
```diff
@@ -1,27 +1,30 @@
     network?: string
 }
 
 @Controller('/api/v1/ideas')
```
We can add `@Tags('deas')` decorator to group the endpoints in swagger

Milszym replied:
Good idea, I have added it :)

aolszewska commented on backend/src/ideas/ideas.controller.ts:
```diff
@@ -1,27 +1,30 @@
+    }
 
     @Get()
     @ApiResponse({
```
There is a short version `@ApiOkResponse`

Milszym replied:
Thanks, I have adjusted that and others Api response descriptions

aolszewska commented on backend/src/ideas/ideas.service.ts:
```diff
@@ -11,43 +13,77 @@ export class IdeasService {
+            throw new NotFoundException('There is no idea with such id')
+        }
+        await this.ideaRepository.save(new Idea(
+            updateIdea.title ?? currentIdea.title,
```
I think it is enough to simply pass `updateIdea.title`. If it is `undefined` we pass `undefined`, if it has a value we pass the value.

Milszym replied:
Hmm but because `title` is required in constructor I can't pass `undefined` there. If I will leave it like this
```
await this.ideaRepository.save(new Idea(
            updateIdea.title,
```
then it will give me an error:
`Argument of type 'string | undefined' is not assignable to parameter of type 'string'.`

aolszewska replied:
Sorry, I need to read the code more carefully :) I didn't notice that the second value is from `currentIdea`, not the `updateIdea`. You can ignore this comment :)

aolszewska commented on backend/src/ideas/ideas.service.ts:
```diff
@@ -11,43 +13,77 @@ export class IdeasService {
+            createIdeaDto.networks.map((network: IdeaNetworkDto) => new IdeaNetwork(
+                network.name,
+                network.value,
+                network.id
```
How do we know the `id` here? For each idea, we create a separate row (at least one, more if more networks added) in the `IdeaNetwork` table, so it should be created upon creating the idea.

Milszym replied:
I have specified `id` in `IdeaNetworkDto` constructor as optional so that whenever we won't pass it, it will got created automatically using `uuid()` method (it's in the constructor implementation).

It's done like this, because it's easier to edit those networks in cascade instead of saving idea first and then networks. 
When we create Idea, we have been passing empty `networks: []` before. But when we want to update `Idea` with networks, then we can't update first every property besides `networks` and then save `networks`, because it would change their `id`s.  

Because of that, I have removed `idea` from `IdeaNetwork` constructor and enabled to pass id. Whenever id is empty (for example because we are just about to create `IdeaNetwork`) we will generate one using `uuid()` method.

Do you have any other idea on how to handle this without?

aolszewska replied:
Discussed during a call, I'll leave this comment opened.

aolszewska replied:
Now that you have changed the `update` method of the service, I think there is no need to pass the id to the constructor of the `IdeaNetwork`. The `BaseEntity` will take care of generating the uuid for the id.

Milszym replied:
You are right, I have removed it.

Milszym commented:
@aolszewska I have added service tests, please check if you think they are sufficient enough for now. I have also adjusted the idea update method in service.

Other thing I have introduced is a `--detectOpenHandles` flag to the debug test npm script. It detects if some connections were not closed. It helped me a lot finding out why tests overlaps with each other. I had a test for `controller` and for `service` checking if search for ideas from particular network is working properly. Once one test wanted to check expected length of result the other already cleared the database. This one was resolved by adding `done` callback to the test like this:
```
it('test', async (done: DoneCallback) => 
[code]
done()
)
```
I think it's quite useful, what do you think about this flag?

aolszewska commented:
The tests look ok, thanks for adding them!
I think it is perfectly ok to add the flag to the `test:debug` command :)
